File: backend/models/Task.js
const mongoose = require("mongoose");

const TaskSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: "User",
    required: true,
    index: true,
  },
  title: { type: String, required: true, trim: true },
  description: { type: String, trim: true },
  status: {
    type: String,
    enum: ["Pending", "In Progress", "Done"],
    default: "Pending",
  },
  deadline: { type: Date, required: true },
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now },
});

TaskSchema.index({ userId: 1, status: 1 });

TaskSchema.pre("findOneAndUpdate", function () {
  this._update.updatedAt = new Date();
});

module.exports = mongoose.model("Task", TaskSchema);

File: mobile/lib/models/task.dart
class Task {
  final String id;
  final String userId;
  final String title;
  final String? description;
  final String status;
  final DateTime deadline;
  final DateTime createdAt;
  final DateTime updatedAt;

  Task({required this.id, required this.userId, required this.title, this.description, required this.status, required this.deadline, required this.createdAt, required this.updatedAt});

  factory Task.fromJson(Map<String, dynamic> json) => Task(
    id: json['_id'],
    userId: json['userId'],
    title: json['title'],
    description: json['description'],
    status: json['status'],
    deadline: DateTime.parse(json['deadline']),
    createdAt: DateTime.parse(json['createdAt']),
    updatedAt: DateTime.parse(json['updatedAt']),
  );

  Map<String, dynamic> toJson() => {
    'title': title,
    'description': description,
    'status': status,
    'deadline': deadline.toIso8601String(),
  };

  Task copyWith({String? title, String? description, String? status, DateTime? deadline}) {
    return Task(
      id: id,
      userId: userId,
      title: title ?? this.title,
      description: description ?? this.description,
      status: status ?? this.status,
      deadline: deadline ?? this.deadline,
      createdAt: createdAt,
      updatedAt: updatedAt,
    );
  }
}
